### PR TITLE
 Take advantage of JSON.NET PropertyNameTable

### PR DESCRIPTION
**Pull request by davidfowl**
- This is a dirty hack to reduce string allocations by taking advantage of the PropertyNameTable on JsonTextReader.
- This is an internal optimization used today by JsonSerializer
so we're borrowing that here.

PS: The performance results aren't great. It looks like it slows things down a bit and the micro benchmarks aren't showing less allocations. dotMemory shows a big difference though.

## https://github.com/aspnet/SignalR/tree/733a3b3c2d911dc0b9585de55fc0fa23af6ca3d0

image

```
            Method |          Input | HubProtocol |       Mean |     Error |     StdDev |      Op/s |  Gen 0 | Allocated |
------------------ |--------------- |------------ |-----------:|----------:|-----------:|----------:|-------:|----------:|
 ReadSingleMessage |   FewArguments |        Json |   2.564 us | 0.0640 us |  0.1805 us | 390,035.0 | 0.0343 |    1088 B |
 ReadSingleMessage | LargeArguments |        Json | 107.738 us | 5.0070 us | 14.0402 us |   9,281.8 | 1.9531 |   59016 B |
 ReadSingleMessage |  ManyArguments |        Json |   4.395 us | 0.1133 us |  0.3252 us | 227,516.6 | 0.0458 |    1648 B |
 ReadSingleMessage |    NoArguments |        Json |   1.556 us | 0.0431 us |  0.1237 us | 642,509.5 | 0.0229 |     768 B |
```

## https://github.com/aspnet/SignalR/tree/davidfowl/json-reader

image

```
           Method |          Input | HubProtocol |       Mean |     Error |     StdDev |    Median |      Op/s |  Gen 0 | Allocated |
------------------ |--------------- |------------ |-----------:|----------:|-----------:|----------:|----------:|-------:|----------:|
 ReadSingleMessage |   FewArguments |        Json |   2.701 us | 0.0937 us |  0.2627 us |  2.625 us | 370,301.2 | 0.0343 |    1088 B |
 ReadSingleMessage | LargeArguments |        Json | 103.064 us | 4.0845 us | 11.6534 us | 99.266 us |   9,702.7 | 1.9531 |   59016 B |
 ReadSingleMessage |  ManyArguments |        Json |   4.573 us | 0.1927 us |  0.5559 us |  4.446 us | 218,662.1 | 0.0458 |    1648 B |
 ReadSingleMessage |    NoArguments |        Json |   1.635 us | 0.0323 us |  0.0659 us |  1.618 us | 611,595.4 | 0.0229 |     768 B |
 ```

1M messages:

## Before

image

## After

image

## Update

I see better results for the latest commit but there's too much variance on my machine to trust. I'll measure more and see if anything shows up.

### CODE REVIEW

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs [hunk 1]
```diff
@@ -117,11 +120,9 @@ private HubMessage ParseMessage(TextReader textReader, IInvocationBinder binder)
                         switch (reader.TokenType)
                         {
                             case JsonToken.PropertyName:
-                                string memberName = reader.Value.ToString();
-
-                                switch (memberName)
                                 {
-                                    case TypePropertyName:
+                                    if (object.ReferenceEquals(reader.Value, TypePropertyName) || string.Equals(reader.Value.ToString(), TypePropertyName))
```
Comment (anurse) [inline src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs]: Gross. I wish there was a way to force `switch` to use reference equality.

Comment (anurse) [reply src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs]: You don't need `object.` here though, right? Can you remove that at least? Maybe wrap this whole thing up in a helper `FastEquals`?

Comment (davidfowl) [reply src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs]: Yea let me do that.

**Comment by JamesNK** [comment]
Not a fan of reflection on internal types.

**Pull request closed without merging**